storage/src/context.rs

```diff
@@ -1,218 +1,73 @@
+    // copies subtree under 'from_key' to new subtree under 'to_key'
+    fn copy_to_diff(&self, context_hash: &Option<ContextHash>, from_key: &ContextKey, to_key: &ContextKey) -> Result<(), ContextError>;
+    // get value for key
+    fn get_key(&self, key: &ContextKey) -> Result<ContextValue, ContextError>;
+    // get values by key prefix
+    fn get_by_key_prefix(&self, prefix: &ContextKey) -> Result<Option<Vec<(ContextKey, ContextValue)>>, ContextError>;
+    // get value for key from a point in history indicated by context hash
+    fn get_key_from_history(&self, context_hash: &ContextHash, key: &ContextKey) -> Result<Option<ContextValue>, ContextError>;
+    // get a list of all key-values under a certain key prefix
+    fn get_key_values_by_prefix(&self, context_hash: &ContextHash, prefix: &ContextKey) -> Result<Option<Vec<(ContextKey, ContextValue)>>, MerkleError>;
+    // convert level number to hash (uses block_storage get_by_block_Level)
+    fn level_to_hash(&self, level: i32) -> Result<ContextHash, ContextError>;
+    // get currently checked out hash
+    fn get_last_commit_hash(&self) -> Option<Vec<u8>>;
+    // get stats from merkle storage
+    fn get_merkle_stats(&self) -> Result<MerkleStorageStats, ContextError>;
 }
 
-/// Stuct which hold diff againts predecessor's context
-pub struct ContextDiff {
-    pub predecessor_index: ContextIndex,
-    pub diff: ContextMap,
-}
-
-impl fmt::Debug for ContextDiff {
-    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        f.debug_struct("ContextDiff")
-            .field("Diff", &self.diff)
-            .finish()
-    }
-}
-
-impl ContextDiff {
-    pub fn new(predecessor_level: Option<usize>, predecessor_context_hash: Option<ContextHash>, diff: ContextMap) -> Self {
-        ContextDiff {
-            predecessor_index: ContextIndex::new(predecessor_level, predecessor_context_hash),
-            diff,
-        }
-    }
-
-    pub fn set(&mut self, context_hash: &Option<ContextHash>, key: &Vec<String>, value: &Vec<u8>) -> Result<(), ContextError> {
-        ensure_eq_context_hash!(context_hash, &self);
-
-        self.diff.insert(to_key(key), Bucket::Exists(value.clone()));
+impl ContextApi for TezedgeContext {
+    fn set(&mut self, _context_hash: &Option<ContextHash>, key: &ContextKey, value: &ContextValue) -> Result<(), ContextError> {
+        let mut merkle = self.merkle.write().expect("lock poisoning");
+        merkle.set(key, value)?;
 
         Ok(())
     }
-}
 
-/// Actual context implementation with context skip list
-pub struct TezedgeContext {
-    block_storage: BlockStorage,
-    storage: ContextList,
-}
+    fn checkout(&self, context_hash: &ContextHash) -> Result<(), ContextError> {
+        let mut merkle = self.merkle.write().expect("lock poisoning");
+        let context_hash_arr: EntryHash = context_hash.as_slice().try_into().expect("EntryHash conversion error");
```

Comment:
@michal-stanek 
`.expect("EntryHash conversion error")` should be better map to ContextError,
I saw `.expect("EntryHash conversion error")` in several places

Reply:
Fixed